language/bytecode-verifier/src/abstract_state.rs

```diff
@@ -129,436 +110,178 @@ impl AbstractState {
     pub fn add_nonce(&mut self, nonce: Nonce) {
-        self.partition.add_nonce(nonce);
+        self.borrow_graph.add_nonce(nonce);
     }
 
-    /// destroys nonce
-    /// borrows of nonce become borrows of any nonce' such that nonce borrows from nonce'
+    /// destroys nonce from borrow graph
     pub fn destroy_nonce(&mut self, nonce: Nonce) {
-        let mut nonce_set = BTreeSet::new();
-        let mut new_locals = BTreeMap::new();
-        let mut new_globals = BTreeMap::new();
-        let mut new_borrows = BTreeMap::new();
-
-        if let Some(borrow_info) = self.borrows.remove(&nonce) {
-            new_borrows = self.strong_propagate(nonce, &borrow_info);
-            match borrow_info {
-                BorrowInfo::BorrowedBy(borrowed_by) => {
-                    nonce_set = borrowed_by.clone();
-                }
-                BorrowInfo::FieldsBorrowedBy(fields_borrowed_by) => {
-                    nonce_set = Self::get_union_of_sets(&BTreeSet::new(), &fields_borrowed_by);
-                }
-            }
-        }
+        self.borrow_graph.destroy_nonce(nonce);
+    }
 
-        for (idx, value) in &self.locals {
-            if let AbstractValue::Value(kind, local_nonce_set) = value {
-                if local_nonce_set.contains(&nonce) {
-                    let mut local_nonce_set_restrict = local_nonce_set.clone();
-                    local_nonce_set_restrict.remove(&nonce);
-                    new_locals.insert(
-                        idx.clone(),
-                        AbstractValue::Value(
-                            *kind,
-                            local_nonce_set_restrict
-                                .union(&nonce_set)
-                                .cloned()
-                                .collect(),
-                        ),
-                    );
-                } else {
-                    new_locals.insert(
-                        idx.clone(),
-                        AbstractValue::Value(*kind, local_nonce_set.clone()),
-                    );
-                }
-            } else {
-                new_locals.insert(idx.clone(), value.clone());
-            }
-        }
+    /// returns all immediate borrows of nonce
+    pub fn borrows(&self, nonce: Nonce) -> BTreeSet<Nonce> {
+        self.borrow_graph.all_borrows(nonce)
+    }
 
-        for (idx, global_nonce_set) in &self.globals {
-            if global_nonce_set.contains(&nonce) {
-                let mut y_restrict = global_nonce_set.clone();
-                y_restrict.remove(&nonce);
-                new_globals.insert(idx.clone(), y_restrict.union(&nonce_set).cloned().collect());
-            } else {
-                new_globals.insert(idx.clone(), global_nonce_set.clone());
-            }
-        }
+    /// returns all borrows of a local
+    pub fn local_borrows(&self, idx: LocalIndex) -> BTreeSet<Nonce> {
+        self.borrow_graph
+            .consistent_borrows(self.frame_root, LabelElem::Local(idx))
```

Comment:
Perhaps explain why the borrow graph uses the term consistent, while the abstract state does not?

Reply:
local_borrows does not exist in abstract_state any more after refactoring.